src/core/mempool.rs

```diff
@@ -675,18 +521,91 @@ impl MemPoolDB {
         block_header_hash: &BlockHeaderHash,
+    ) -> Result<usize, db_error> {
+        let sql = "SELECT * FROM mempool WHERE consensus_hash = ?1 AND block_header_hash = ?2";
+        let args: &[&dyn ToSql] = &[consensus_hash, block_header_hash];
+        let rows = query_rows::<MemPoolTxInfo, _>(conn, &sql, args)?;
+        Ok(rows.len())
+    }
+
+    /// Get the next nonce/timestamp after this nonce that occurs in the height range.
+    pub fn get_next_nonce(
+        conn: &DBConn,
+        min_height: Option<u64>,
+        max_height: u64,
+        nonce: Option<u64>,
+    ) -> Result<Option<(u64, u64)>, db_error> {
+        let nonce = match nonce {
+            None => -1,
+            Some(n) => u64_to_sql(n)?,
+        };
+        let min_height_sql_arg = match min_height {
+            None => -1,
+            Some(h) => u64_to_sql(h)?,
+        };
+        let sql = "SELECT origin_nonce, accept_time FROM mempool WHERE origin_nonce > ?1 AND \
+        height > ?2 AND height <= ?3 ORDER BY origin_nonce, accept_time ASC LIMIT 1";
+
+        let args: &[&dyn ToSql] = &[&nonce, &min_height_sql_arg, &u64_to_sql(max_height)?];
+        query_row(conn, sql, args)
+    }
+
+    /// Get the next timestamp after this one that occurs with this nonce in the height range.
+    pub fn get_next_timestamp_at_nonce(
+        conn: &DBConn,
+        min_height: Option<u64>,
+        max_height: u64,
         timestamp: u64,
-    ) -> Result<Option<u64>, db_error> {
-        let sql = "SELECT accept_time FROM mempool WHERE accept_time > ?1 AND consensus_hash = ?2 AND block_header_hash = ?3 ORDER BY accept_time ASC LIMIT 1";
-        let args: &[&dyn ToSql] = &[&u64_to_sql(timestamp)?, consensus_hash, block_header_hash];
+        nonce: u64,
+    ) -> Result<Option<(u64)>, db_error> {
+        let min_height_sql_arg = match min_height {
+            None => -1,
+            Some(h) => u64_to_sql(h)?,
+        };
+
+        let sql = "SELECT accept_time FROM mempool WHERE accept_time > ?1 AND \
+        origin_nonce = ?2 AND height > ?3 AND height <= ?4 ORDER BY accept_time ASC LIMIT 1";
+        let args: &[&dyn ToSql] = &[
+            &u64_to_sql(timestamp)?,
+            &u64_to_sql(nonce)?,
+            &min_height_sql_arg,
+            &u64_to_sql(max_height)?,
+        ];
         query_row(conn, sql, args)
     }
 
+    /// Get all transactions at a particular nonce and timestamp on a given chain tip.
+    /// Order them by sponsor nonce.
+    pub fn get_txs_at_nonce_and_timestamp(
+        conn: &DBConn,
+        min_height: Option<u64>,
+        max_height: u64,
+        nonce: u64,
+        timestamp: u64,
+    ) -> Result<Vec<MemPoolTxInfo>, db_error> {
+        let min_height_sql_arg = match min_height {
+            None => -1,
+            Some(h) => u64_to_sql(h)?,
+        };
+
+        let sql = "SELECT * FROM mempool WHERE origin_nonce = ?1 AND accept_time = ?2 AND \
+        height > ?3 AND height <= ?4 ORDER BY sponsor_nonce ASC";
```

Comment:
What happens if there are a _lot_ of transactions with the same nonce and timestamp?  Should there be some kind of pagination logic here to prevent this from becoming a DDoS vector?  If not, then perhaps some kind of upstream insert rate-limiting should be in place to prevent this list from getting too large.

Reply:
Yeah, I can definitely add some type of pagination logic.